Document node discovery annotations (#40462)
kubernetes/website: content/en/docs/reference/labels-annotations-taints/_index.md

Author: > Should we also document `feature.node.kubernetes.io/*` ?

There are many Labels for features , Reference is here: https://kubernetes-sigs.github.io/node-feature-discovery/v0.12/usage/features.html

should we add all of them? 
Won't it be better to add them in other PR or mentioning this reference with note should be enough because these are only applicable when NFD is running on node.

Reviewer: If we document the annotations it would probably make sense to document this as well. Not document all the feature labels but mention that labels prefixed with`feature.node.kubernetes.io/` are (likely) from NFD

Author: I agree how would you like to see it done . should i add this like :

```
### feature.node.kubernetes.io/*
Used on: Nodes 
Example: `feature.node.kubernetes.io/network-sriov.capable: "true"`

These features are advertised as labels in the Kubernetes Node object on nodes running NFD.
All built-in labels use the feature.node.kubernetes.io label namespace and have the following  format.:
`feature.node.kubernetes.io/<feature-name>: <true>`
NFD has many extension points for creating vendor and application-specific labels. See the [customization guide](https://kubernetes-sigs.github.io/node-feature-discovery/v0.12/usage/customization-guide) for detailed documentation.
```

Reviewer 2: Please try that.

Author: Hi @sftim, thanks for the response. I'll add it .
Can we do this in general way? like putting `### feature.node.kubernetes.io/*`

Reviewer 2: Sure

If a Node is in a cloud provider specified shutdown state, the Node gets tainted accordingly with `node.cloudprovider.kubernetes.io/shutdown` and the taint effect of `NoSchedule`.

{{< note >}}
Node Feature Discovery(NFD) is a Kubernetes add-on for detecting hardware features and

> system configuration. To learn more about NFD and its components go to its official

Reviewer 3: ```suggestion
system configuration. To learn more about NFD and its components, visit its official
```

[documentation](https://kubernetes-sigs.github.io/node-feature-discovery/v0.7/get-started/).

> {{< /note >}}

Author: So , notes are mentioned below the related labels. 
I didn't know it. Will take care of the same.
I'll explain NFD in each entry but then there wont be much need of NOTE for NFD. WDYT?

### nfd.node.kubernetes.io/master.version

Example: `nfd.node.kubernetes.io/master.version: v0.6.0`

This annotation defines version of the NFD-Master instance running on the node.

> It is used for informative use only.

Reviewer 3: It'd be better if we clarify who is using it.

Author: Hi @divya-mohan0209 , these annotations are under the list where it's mentioned that:
`The taints listed below are always used on Nodes
`
it's [here](https://kubernetes.io/docs/reference/labels-annotations-taints/#node-kubernetes-io-not-ready)

Reviewer 3: I meant that in the sentence you've mentioned that these annotations are used for informative purposes only. It'd help clarify what is using these annotations. Is it NFD that's using them?

Author: The master.version is only created on nodes running nfd-master. so yes it is used for NFD mainly.

Author: I can mention to each , 
`Used on: Nodes running NFD`
WDYT @divya-mohan0209 ?

Reviewer 2: `Used on: Nodes`

is right. In the text, mention that it only applies for nodes where NFD is running.

> This annotation defines version of the NFD-Master instance running on the node.
> It is used for informative use only.

Reviewer 2: ```suggestion
For node(s) where the Node Feature Discovery (NFD)
[master](https://kubernetes-sigs.github.io/node-feature-discovery/stable/usage/nfd-master.html)
is scheduled, this annotation records the version of the NFD master.
```

Author: Done . PTAL @sftim @divya-mohan0209

### nfd.node.kubernetes.io/worker.version

Example: `nfd.node.kubernetes.io/worker.version: v0.4.0`

This annotation defines version of the NFD-Worker instance running on the node.
It's used for informative use only.

### nfd.node.kubernetes.io/feature-labels

> Example: `nfd.node.kubernetes.io/feature-labels: cpu-cpuid.ADX,cpu-cpuid.AESNI,cpu-hardware_multithreading,kernel-version.full,system-os_release.ID ...`

Reviewer 2: ```suggestion
Example: `nfd.node.kubernetes.io/feature-labels: "cpu-cpuid.ADX,cpu-cpuid.AESNI,cpu-hardware_multithreading,kernel-version.full"`
```

Example: `nfd.node.kubernetes.io/extended-resources: sgx.intel.com/epc`

This annotation defines a comma-separated list of node extended resources managed by NFD.
NFD uses this internally so it must not be edited by users.

> {{< note >}}

Reviewer 3: Adding this at the beginning makes more sense, IMO, since we are referencing nfd multiple times.

Author: I have mentioned it in beginning earlier as a [preview](https://github.com/kubernetes/website/issues/31612#issuecomment-1491997100) but later, as per the doc all notes are after the statements ,I added it in end.

I agree it will give more info.

Author: @divya-mohan0209 any other information that i can add in note ? or is just good enough as it is?

> This annotation defines comma-separated list of node labels managed by NFD.
> NFD uses this internally so it must not be edited by users.

Reviewer 2: ```suggestion
This annotation records a comma-separated list of node labels managed by
[Node Feature Discovery](https://kubernetes-sigs.github.io/node-feature-discovery/) (NFD).
NFD uses this for an internal mechanism; you should not edit this annotation yourself.
```

Example: `nfd.node.kubernetes.io/extended-resources: sgx.intel.com/epc`

> This annotation defines a comma-separated list of node extended resources managed by NFD.
> NFD uses this internally so it must not be edited by users.

Reviewer 2: ```suggestion
This annotation records a comma-separated list of [extended resources](/docs/concepts/configuration/manage-resources-containers/#extended-resources) managed by
[Node Feature Discovery](https://kubernetes-sigs.github.io/node-feature-discovery/) (NFD).
NFD uses this for an internal mechanism; you should not edit this annotation yourself.
```

Example: `nfd.node.kubernetes.io/feature-labels: "cpu-cpuid.ADX,cpu-cpuid.AESNI,cpu-hardware_multithreading,kernel-version.full"`

Used on: Nodes

This annotation records a comma-separated list of node labels managed by
[Node Feature Discovery](https://kubernetes-sigs.github.io/node-feature-discovery/) (NFD).
NFD uses this for an internal mechanism. You should not edit this annotation yourself.

### nfd.node.kubernetes.io/extended-resources

Example: `nfd.node.kubernetes.io/extended-resources: "accelerator.acme.example/q500,example.com/coprocessor-fx5"`

Used on: Nodes

This annotation records a comma-separated list of [extended resources](/docs/concepts/configuration/manage-resources-containers/#extended-resources) managed by
[Node Feature Discovery](https://kubernetes-sigs.github.io/node-feature-discovery/) (NFD).
NFD uses this for an internal mechanism. You should not edit this annotation yourself.

{{< note >}}
These annotations only applies for nodes where NFD is running.
Node Feature Discovery(NFD) is a Kubernetes add-on for detecting hardware features and
system configuration.
To learn more about NFD and its components go to its official

> [documentation](https://kubernetes-sigs.github.io/node-feature-discovery/v0.7/get-started/).

Author: That seems like a right way to keep it in sync. Thanks for the review @marquiz

If a Node is in a cloud provider specified shutdown state, the Node gets tainted accordingly with `node.cloudprovider.kubernetes.io/shutdown` and the taint effect of `NoSchedule`.

### nfd.node.kubernetes.io/master.version

Example: `nfd.node.kubernetes.io/master.version: "v0.6.0"`

Used on: Nodes

For node(s) where the Node Feature Discovery (NFD)
[master](https://kubernetes-sigs.github.io/node-feature-discovery/stable/usage/nfd-master.html)
is scheduled, this annotation records the version of the NFD master.
It is used for informative use only.

### nfd.node.kubernetes.io/worker.version

Example: `nfd.node.kubernetes.io/worker.version: "v0.4.0"`

Used on: Nodes

This annotation records the version for a Node Feature Discovery's
[worker](https://kubernetes-sigs.github.io/node-feature-discovery/stable/usage/nfd-worker.html)
if there is one running on a node.
It's used for informative use only.

### nfd.node.kubernetes.io/feature-labels

Example: `nfd.node.kubernetes.io/feature-labels: "cpu-cpuid.ADX,cpu-cpuid.AESNI,cpu-hardware_multithreading,kernel-version.full"`

Used on: Nodes

> This annotation records a comma-separated list of node labels managed by

Reviewer 2: ```suggestion
This annotation records a comma-separated list of node feature labels managed by
```

If a Node is in a cloud provider specified shutdown state, the Node gets tainted accordingly with `node.cloudprovider.kubernetes.io/shutdown` and the taint effect of `NoSchedule`.

### feature.node.kubernetes.io/*
Used on: Nodes 
Example: `feature.node.kubernetes.io/network-sriov.capable: "true"`

These features are advertised as labels in the Kubernetes Node object on nodes running NFD.

> All built-in labels use the feature.node.kubernetes.io label namespace and have the following  format.:
> `feature.node.kubernetes.io/<feature-name>: <true>`
> NFD has many extension points for creating vendor and application-specific labels. See the [customization guide](https://kubernetes-sigs.github.io/node-feature-discovery/v0.12/usage/customization-guide) for detailed documentation.

Reviewer 4: ```suggestion
All built-in labels use the feature.node.kubernetes.io label namespace and have the format
`feature.node.kubernetes.io/<feature-name>: <true>`.
NFD has many extension points for creating vendor and application-specific labels.
For details, see the [customization guide](https://kubernetes-sigs.github.io/node-feature-discovery/v0.12/usage/customization-guide).
```

[master](https://kubernetes-sigs.github.io/node-feature-discovery/stable/usage/nfd-master.html)
is scheduled, this annotation records the version of the NFD master.
It is used for informative use only.

### nfd.node.kubernetes.io/worker.version

Example: `nfd.node.kubernetes.io/worker.version: "v0.4.0"`

Used on: Nodes

This annotation records the version for a Node Feature Discovery's
[worker](https://kubernetes-sigs.github.io/node-feature-discovery/stable/usage/nfd-worker.html)
if there is one running on a node.
It's used for informative use only.

### nfd.node.kubernetes.io/feature-labels

Example: `nfd.node.kubernetes.io/feature-labels: "cpu-cpuid.ADX,cpu-cpuid.AESNI,cpu-hardware_multithreading,kernel-version.full"`

Used on: Nodes

This annotation records a comma-separated list of node feature labels managed by
[Node Feature Discovery](https://kubernetes-sigs.github.io/node-feature-discovery/) (NFD).
NFD uses this for an internal mechanism. You should not edit this annotation yourself.

### nfd.node.kubernetes.io/extended-resources

Example: `nfd.node.kubernetes.io/extended-resources: "accelerator.acme.example/q500,example.com/coprocessor-fx5"`

Used on: Nodes

> This annotation records a comma-separated list of [extended resources](/docs/concepts/configuration/manage-resources-containers/#extended-resources) managed by

Reviewer 4: ```suggestion
This annotation records a comma-separated list of [extended resources](/docs/concepts/configuration/manage-resources-containers/#extended-resources)
managed by
```

[Node Feature Discovery](https://kubernetes-sigs.github.io/node-feature-discovery/) (NFD).
NFD uses this for an internal mechanism. You should not edit this annotation yourself.

{{< note >}}

> These annotations only applies for nodes where NFD is running.

Reviewer 4: ```suggestion
These annotations only applies to nodes where NFD is running.
```

> Node Feature Discovery(NFD) is a Kubernetes add-on for detecting hardware features and
> system configuration.

Reviewer 4: ```suggestion
```
